In [ ]:
import pandas as pd
import requests
import socket
import ipaddress

In [ ]:
# prompt: read sample.txt and find and replace ":::" and "::" with ":"

import pandas as pd

def find_and_replace(input_file, output_file):
    try:
        with open(input_file, 'r') as infile, open(output_file, 'w') as outfile:
            for line in infile:
                # Replace ":::" and "::" with ":"
                modified_line = line.replace(":::", ":").replace("::", ":")
                outfile.write(modified_line)
    except FileNotFoundError:
        print(f"Error: Input file '{input_file}' not found.")
    except Exception as e:
        print(f"An error occurred: {e}")

# Example usage
input_filename = "sample.txt"  # Replace with your input file name
output_filename = "mod.txt"  # Replace with your desired output file name
find_and_replace(input_filename, output_filename)

In [ ]:
# prompt: for lines in modified_sample.txt with 4 or more colons, remove everything including and after the fourth colon

import pandas as pd


def find_and_replace(input_file, output_file):
    try:
        with open(input_file, 'r') as infile, open(output_file, 'w') as outfile:
            for line in infile:
                # Replace ":::" and "::" with ":"
                modified_line = line.replace(":::", ":").replace("::", ":")
                outfile.write(modified_line)
    except FileNotFoundError:
        print(f"Error: Input file '{input_file}' not found.")
    except Exception as e:
        print(f"An error occurred: {e}")


def process_modified_file(input_file, output_file):
    try:
        with open(input_file, 'r') as infile, open(output_file, 'w') as outfile:
            for line in infile:
                colons = line.count(':')
                if colons >= 4:
                    parts = line.split(':', 4)  # Split at most 4 times
                    outfile.write(':'.join(parts[:4]) + '\n')
                else:
                    outfile.write(line)
    except FileNotFoundError:
        print(f"Error: Input file '{input_file}' not found.")
    except Exception as e:
        print(f"An error occurred: {e}")


# Example usage
input_filename = "sample.txt"  # Replace with your input file name
modified_filename = "modified_sample.txt"  # Replace with your desired output file name
find_and_replace(input_filename, modified_filename)

final_output_filename = "final_output.txt"
process_modified_file(modified_filename, final_output_filename)

In [ ]:
# prompt: use pd read csv and print the tail

import pandas as pd

# Assuming 'final_output.txt' is your file
try:
    df = pd.read_csv('final_output.txt', sep=':', header=None, engine='python')
    print(df.tail())
except FileNotFoundError:
    print("Error: 'final_output.txt' not found.")
except pd.errors.ParserError:
    print("Error: Could not parse 'final_output.txt'. Check file format and delimiter.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

            0                                                  1       2  \
364072  https  //www.orange.eg/ar/business/myaccount/captcha-...  ******   
364073  https                                   //www.roblox.com  ******   
364074  https           //portal.nttrv2.helixsense.com/web/login  ******   
364075  https                               //www.sammobile.com/  ******   
364076  https                                        //btcs.love  ******   

               3  
364072  ********  
364073  ********  
364074  ********  
364075  ********  
364076  ********  


In [ ]:
# prompt: append ":" + column 1 to column 0, and then shift columns 2 and 3 over

import pandas as pd

# Assuming 'final_output.txt' is your file
try:
    df = pd.read_csv('final_output.txt', sep=':', header=None, engine='python')

    # Append ":" + column 1 to column 0
    df[0] = df[0] + ":" + df[1].astype(str)

    # Shift columns 2 and 3 over (assuming you want to delete the original column 1)
    df = df.drop(columns=1)

    # Rename columns (optional, adjust as needed)
    df.columns = ['URI', 'username', 'password']

    print(df.tail())

except FileNotFoundError:
    print("Error: 'final_output.txt' not found.")
except pd.errors.ParserError:
    print("Error: Could not parse 'final_output.txt'. Check file format and delimiter.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

                                                      URI username  password
364072  https://www.orange.eg/ar/business/myaccount/ca...   ******  ********
364073                             https://www.roblox.com   ******  ********
364074     https://portal.nttrv2.helixsense.com/web/login   ******  ********
364075                         https://www.sammobile.com/   ******  ********
364076                                  https://btcs.love   ******  ********


In [ ]:
# prompt: output this df to "final_output_final.txt"

try:
    df.to_csv("final_output_final.txt", sep=":", header=None, index=False)
    print(f"DataFrame successfully saved to final_output_final.txt")
except Exception as e:
    print(f"An error occurred while saving the DataFrame: {e}")

DataFrame successfully saved to final_output_final.txt


In [ ]:
# prompt: update the df with a column on the left 'id' and print the tail and save it as "final_output_final_final.txt"

import pandas as pd

try:
    df = pd.read_csv('final_output_final.txt', sep=':', header=None, engine='python')

    # Insert 'id' column as the first column with values from 0 to n-1
    df.insert(0, 'id', range(len(df)))

    print(df.tail())

    df.to_csv("final_output_final_final.txt", sep=":", header=None, index=False)
    print(f"DataFrame successfully saved to final_output_final_final.txt")

except FileNotFoundError:
    print("Error: 'final_output.txt' not found.")
except pd.errors.ParserError:
    print("Error: Could not parse 'final_output.txt'. Check file format and delimiter.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

            id                                                  0       1  \
364072  364072  https://www.orange.eg/ar/business/myaccount/ca...  ******   
364073  364073                             https://www.roblox.com  ******   
364074  364074     https://portal.nttrv2.helixsense.com/web/login  ******   
364075  364075                         https://www.sammobile.com/  ******   
364076  364076                                  https://btcs.love  ******   

               2  
364072  ********  
364073  ********  
364074  ********  
364075  ********  
364076  ********  
DataFrame successfully saved to final_output_final_final.txt


In [ ]:
# prompt: take final_output_final_final.txt, use ipaddress to find out if the uri is a private ip. update the df and save that as a new collumn. save it as final_output_with_priv.txt

import pandas as pd
import ipaddress
from urllib.parse import urlparse


def is_private_ip(uri_str):
    try:
        parsed_uri = urlparse(uri_str)
        # Extract hostname (which might be an IP)
        hostname = parsed_uri.hostname
        # Check if hostname is a valid IP and private
        ip = ipaddress.ip_address(hostname)
        return ip.is_private
    except (ValueError, TypeError):  # Handle invalid IPs or missing hostnames
        return False



try:
    df = pd.read_csv('final_output_final_final.txt', sep=':', header=None, engine='python')



    # Apply the function to the URI column and create a new column 'is_private'
    df['is_private'] = df[1].apply(is_private_ip)

    print(df.tail())


    print(df[df['is_private'] == True].tail())



    df.to_csv("final_output_with_priv.txt", sep=":", header=None, index=False)
    print(f"DataFrame successfully saved to final_output_with_priv.txt")

except FileNotFoundError:
    print("Error: 'final_output_final_final.txt' not found.")
except pd.errors.ParserError:
    print("Error: Could not parse 'final_output_final_final.txt'. Check file format and delimiter.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

             0                                                  1       2  \
364072  364072  https://www.orange.eg/ar/business/myaccount/ca...  ******   
364073  364073                             https://www.roblox.com  ******   
364074  364074     https://portal.nttrv2.helixsense.com/web/login  ******   
364075  364075                         https://www.sammobile.com/  ******   
364076  364076                                  https://btcs.love  ******   

               3  is_private  
364072  ********       False  
364073  ********       False  
364074  ********       False  
364075  ********       False  
364076  ********       False  
             0                                         1       2         3  \
363270  363270  http://192.168.0.1/device-map/router.asp  ******  ********   
363310  363310                       http://192.168.0.1/  ******  ********   
363615  363615                      https://192.168.1.1/  ******  ********   
363716  363716                    https

1. ! A user should be able to find all accounts associated with a specific domain or IP address.
2. x A user should be able to find all accounts associated with a specific application (e.g., wordpress, citrix).
3. x A user should be able to find all accounts associated with a specific port (e.g., 443, 3000, 8080).
4. ? A user should be able to find all accounts associated with a specific URL path (e.g., /rdweb/, wp-login.php, /citrix/).
5. ! A user should be able to exclude non-routable IP ranges from searches (e.g., 127.0.0.1, localhost, 192.168.x.x).
6. x A user should be able to search for or exclude specific tags (e.g., resolved, local-ip, parked, active).
7. x Confirm that a domain resolves to an IP address. If not, tag the domain as "unresolved."
8. x Confirm that the URL is accessible (status code 200), and tag the URL accordingly.
9. x Retrieve and store the title of the URL in the database.
10. x Assess if a domain is parked or has no content, and tag the domain accordingly.
11. x Identify whether the URL contains a login form, and tag the URL if applicable.
12. x Identify the web application associated with the login form (e.g., WordPress, Citrix, Cisco, Coremail, RDWeb), and tag the URL accordingly.
13. x Assess whether the login form requires a CAPTCHA, and tag the URL if applicable.
14. x Assess whether the login form requires an OTP (multi-factor authentication), and tag the URL if applicable.
15. x Assess whether the domain has ever been ransomed or is part of a breach, and tag the URL if applicable.


out of scope :(

In [ ]:

#tuf

class Domain:
  def __init__(self, domain):
      self.domain = domain
      self.resolve = self.resolve()
      self.is_parked = self.is_domain_parked(domain)

  def resolve(self):
      try:
          ip_address = socket.gethostbyname(self.domain)
          return ip_address
      except socket.gaierror:
          return None

  def has_ip(url):
      try:
          ip_address = socket.gethostbyname(url)
          return True
      except socket.gaierror:
          return False

  def has_ip(url):
      try:
           socket.gethostbyname(url)
      except socket.gaierror:
          return None

  def is_domain_parked(domain):
      try:
          response = requests.get(f"http://{domain}", timeout=10)
          if response.status_code == 200:
              # Look for common parking keywords
              if "domain is parked" in response.text.lower() or "this domain is for sale" in response.text.lower():
                  return True
          return False
      except requests.RequestException:
          return False


In [ ]:
class URI:
  def __init__(self, uri):
      self.uri = uri
      self.uri_acessible = self.uri_acessible()
      self.has_ip = self.has_ip()
      self.ip_local = self.is_private_ip()
      self.resolve = self.resolve()
      self.title = self.get_title()
      self.has_login_form = self.has_login_form()
      self.get_web_application = self.get_web_application()
      self.login_requires_captcha = self.login_requires_captcha()

  def resolve(self):
      try:
          ip_address = socket.gethostbyname(self.uri)
          return ip_address
      except socket.gaierror:
          return None

  def uri_acessible(uri):
      try:
          response = requests.get(uri)
          if response.status_code == 200:
              return True
          else:
              return False
      except requests.exceptions.RequestException:
          return False


  def has_ip(uri):
      try:
          ip_address = socket.gethostbyname(uri)
          return True
      except socket.gaierror:
          return False

  def get_title(self):
      try:
          response = requests.get(self.uri)
          soup = BeautifulSoup(response.text, 'html.parser')
          title = soup.title.string.strip()
          return title
      except requests.exceptions.RequestException:
          return None


  def has_login_form(self):
      try:
          response = requests.get(self.uri)
          soup = BeautifulSoup(response.text, 'html.parser')
          login_form = soup.find('form', {'action': '/login'})
          return bool(login_form)
      except requests.exceptions.RequestException:
          return False


  def login_requires_captcha(self):
      try:
          response = requests.get(self.uri)
          soup = BeautifulSoup(response.text, 'html.parser')
          captcha_form = soup.find('form', {'action': '/login'})
          if captcha_form:
              captcha_field = captcha_form.find('input', {'name': 'captcha'})
              return bool(captcha_field)
          return False
      except requests.exceptions.RequestException:
          return False

  def login_requires_otp(self):
      try:
          response = requests.get(self.uri)
          soup = BeautifulSoup(response.text, 'html.parser')
          otp_form = soup.find('form', {'action': '/login'})
          if otp_form:
              otp_field = otp_form.find('input', {'name': 'otp'})
              return bool(otp_field)
          return False
      except requests.exceptions.RequestException:
          return False

  def get_web_application(self):
      try:
          response = requests.get(self.uri)
          soup = BeautifulSoup(response.text, 'html.parser')
          login_form = soup.find('form', {'action': '/login'})
          if login_form:
              action = login_form.get('action', '').lower()
              if 'wordpress' in action:
                  return 'WordPress'
              elif 'citrix' in action:
                  return 'Citrix'
              elif 'cisco' in action:
                  return 'Cisco'
              elif 'coremail' in action:
                  return 'Coremail'
              elif 'rdweb' in action:
                  return 'RDWeb'
          return None
      except requests.exceptions.RequestException:
          return None

      def is_private_ip(ip_address):
          """Checks if an IP address is private."""
          try:
              ip = ipaddress.ip_address(ip_address)
              return ip.is_private
          except ValueError:
              return False  # Invalid IP address


